### PR TITLE
Mesh 1569/pius v2 integration

### PR DESCRIPTION
**Pull request by fmiguelgarcia**
Some changes to perform the integration with Polymesh Core:
- It uses Substrate from `crate.io`, which allows Core to override deps easily.
- New wrappers to `Encode` and `Decode` external types like `Scaler` or `RisttretoPoint`.
- Some types now supports `Encode` & `Decode`

### CODE REVIEW

**Comment by Arash-Afshar** [comment]
@fmiguelgarcia why do we need new wrappers for dalek types (i.e, what was wrong with the previous wrapper)? Also, if it is pius related, it is probably best to submit the PR to the private repo.

**Comment by fmiguelgarcia** [comment]
> @fmiguelgarcia why do we need new wrappers for dalek types (i.e, what was wrong with the previous wrapper)? Also, if it is pius related, it is probably best to submit the PR to the private repo.

This new implementation is less error prone, and also abstracts the impl of decode/encode of `Scale` and `RistrettoPoint`.
A part of that improvement, It also contains changes needed by core to use the latest version of the cryptography library.


att @Arash-Afshar

**Review by Arash-Afshar — APPROVED** [review]
lgtm

File: cryptography-core/src/cdd_claim/cdd_claim_data.rs [hunk 1]
```diff
@@ -37,22 +36,20 @@ impl CddClaimData {
...
 
 impl Decode for CddClaimData {
     fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
-        let investor_did = <[u8; SCALAR_SIZE]>::decode(input)?;
-        let investor_did = Scalar::from_bits(investor_did);
-
-        let investor_unique_id = <[u8; SCALAR_SIZE]>::decode(input)?;
-        let investor_unique_id = Scalar::from_bits(investor_unique_id);
+        let investor_did = <ScalarDecoder>::decode(input)?.0;
+        let investor_unique_id = <ScalarDecoder>::decode(input)?.0;
```
Comment (vkomenda) [inline cryptography-core/src/cdd_claim/cdd_claim_data.rs:52]: These could be abbreviated with a closure as well.

**Review by vkomenda — COMMENTED** [review]
I haven't checked encoders or decoders but overall everything looks great.

File: cryptography-core/src/asset_proofs/ciphertext_refreshment_proof.rs [hunk 2]
```diff
@@ -88,26 +85,19 @@ impl Default for CipherTextRefreshmentInitialMessage {
...
-        let a = CompressedRistretto(a).decompress().ok_or_else(|| {
-            CodecError::from("CipherTextRefreshmentInitialMessage::a point is invalid")
-        })?;
-        let b = CompressedRistretto(b).decompress().ok_or_else(|| {
-            CodecError::from("CipherTextRefreshmentInitialMessage::b point is invalid")
-        })?;
+        let a = <RistrettoPointDecoder>::decode(input)?.0;
+        let b = <RistrettoPointDecoder>::decode(input)?.0;
 
         Ok(CipherTextRefreshmentInitialMessage { a, b })
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply cryptography-core/src/asset_proofs/ciphertext_refreshment_proof.rs:102]: It works adding the field name on the struct initialization, and making `f` as `mut`. 
The problem here is the _side effect_ of mutable `f` and the evaluation order.  After review [this](https://internals.rust-lang.org/t/rust-expression-order-of-evaluation/2605), and [this](https://doc.rust-lang.org/nightly/reference/expressions.html?highlight=evaluation%20order), the evaluation order (left-to-right) is defined in the specification and it should not change. So far so good.

However it seems a good idea to do any order-dependent operation in a separated statement.

File: cryptography-core/src/asset_proofs/elgamal_encryption.rs [hunk 3]
```diff
@@ -97,27 +91,19 @@ pub struct CipherText {
...
-        let x = CompressedRistretto(x)
-            .decompress()
-            .ok_or_else(|| CodecError::from("CipherText X point is invalid"))?;
-        let y = CompressedRistretto(y)
-            .decompress()
-            .ok_or_else(|| CodecError::from("CipherText Y point is invalid"))?;
+        let x = <RistrettoPointDecoder>::decode(input)?.0;
+        let y = <RistrettoPointDecoder>::decode(input)?.0;
 
         Ok(CipherText { x, y })
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply cryptography-core/src/asset_proofs/elgamal_encryption.rs:108]: See previous response.

File: cryptography-core/src/asset_proofs/encrypting_same_value_proof.rs [hunk 4]
```diff
@@ -79,30 +78,23 @@ pub struct EncryptingSameValueInitialMessage {
...
-            .decompress()
-            .ok_or_else(|| CodecError::from("EncryptingSameValueInitialMessage `a2` is invalid"))?;
-        let b = CompressedRistretto(b)
-            .decompress()
-            .ok_or_else(|| CodecError::from("EncryptingSameValueInitialMessage `b` is invalid"))?;
+        let a1 = <RistrettoPointDecoder>::decode(input)?.0;
+        let a2 = <RistrettoPointDecoder>::decode(input)?.0;
+        let b = <RistrettoPointDecoder>::decode(input)?.0;
 
         Ok(EncryptingSameValueInitialMessage { a1, a2, b })
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply cryptography-core/src/asset_proofs/encrypting_same_value_proof.rs:99]: See previous response.

File: cryptography-core/src/asset_proofs/one_out_of_many_proof.rs [hunk 5]
```diff
@@ -291,37 +291,25 @@ impl R1ProofInitialMessage {
...
-            .map(|compressed| {
-                CompressedRistretto(*compressed)
-                    .decompress()
-                    .ok_or_else(|| CodecError::from("R1ProofInitialMessage is invalid"))
-            })
-            .collect::<Result<Vec<_>, _>>()?;
+        let a = <RistrettoPointDecoder>::decode(input)?.0;
+        let b = <RistrettoPointDecoder>::decode(input)?.0;
+        let c = <RistrettoPointDecoder>::decode(input)?.0;
+        let d = <RistrettoPointDecoder>::decode(input)?.0;
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply cryptography-core/src/asset_proofs/one_out_of_many_proof.rs:310]: See previous response.

File: cryptography-core/src/cdd_claim/cdd_claim_data.rs (hunk 1 above)
Comment (fmiguelgarcia) [reply cryptography-core/src/cdd_claim/cdd_claim_data.rs:52]: See previous response.

File: cryptography-core/src/asset_proofs/encrypting_same_value_proof.rs [hunk 6]
```diff
@@ -48,21 +49,19 @@ pub struct EncryptingSameValueFinalResponse {
...
     fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
-        let (z1, z2) = <([u8; 32], [u8; 32])>::decode(input)?;
-        let z1 = Scalar::from_canonical_bytes(z1)
-            .ok_or_else(|| CodecError::from("EncryptingSameValueFinalResponse `z1` is invalid"))?;
-        let z2 = Scalar::from_canonical_bytes(z2)
-            .ok_or_else(|| CodecError::from("EncryptingSameValueFinalResponse `z2` is invalid"))?;
+        let z1 = <ScalarDecoder>::decode(input)?.0;
+        let z2 = <ScalarDecoder>::decode(input)?.0;
 
         Ok(EncryptingSameValueFinalResponse { z1, z2 })
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply cryptography-core/src/asset_proofs/encrypting_same_value_proof.rs:66]: See previous response.

File: cryptography-core/src/asset_proofs/elgamal_encryption.rs [hunk 7]
```diff
@@ -62,25 +62,19 @@ impl From<(Scalar, &mut StdRng)> for CommitmentWitness {
...
     fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
-        let (value, blinding) = <([u8; 32], [u8; 32])>::decode(input)?;
-        let value = Scalar::from_canonical_bytes(value)
-            .ok_or_else(|| CodecError::from("CommitmentWitness value point is invalid"))?;
-        let blinding = Scalar::from_canonical_bytes(blinding)
-            .ok_or_else(|| CodecError::from("CommitmentWitness blinding point is invalid"))?;
+        let value = <ScalarDecoder>::decode(input)?.0;
+        let blinding = <ScalarDecoder>::decode(input)?.0;
 
         Ok(CommitmentWitness { value, blinding })
```
[the remark this answers is not in the document]
Comment (fmiguelgarcia) [reply cryptography-core/src/asset_proofs/elgamal_encryption.rs:79]: See previous response.

**Pull request merged**